googlecloud-plugins/src/main/java/co/cask/hydrator/plugin/sink/GCSBatchSink.java

```diff
@@ -0,0 +1,152 @@
+import co.cask.cdap.api.annotation.Macro;
+import co.cask.cdap.api.annotation.Name;
+import co.cask.cdap.api.data.batch.Output;
+import co.cask.cdap.api.data.batch.OutputFormatProvider;
+import co.cask.cdap.api.data.format.StructuredRecord;
+import co.cask.cdap.etl.api.batch.BatchSinkContext;
+import co.cask.hydrator.common.ReferenceBatchSink;
+import co.cask.hydrator.common.ReferencePluginConfig;
+import co.cask.hydrator.common.batch.sink.SinkOutputFormatProvider;
+import com.google.common.annotations.VisibleForTesting;
+import com.google.common.reflect.TypeToken;
+import com.google.gson.Gson;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import java.lang.reflect.Type;
+import java.util.HashMap;
+import java.util.Map;
+import javax.annotation.Nullable;
+
+/**
+ * {@link GCSBatchSink} that stores the data to Google Cloud Storage Bucket.
+ * @param <KEY_OUT> the type of key the sink outputs
+ * @param <VAL_OUT> the type of value the sink outputs
+ */
+
+public abstract class GCSBatchSink<KEY_OUT, VAL_OUT> extends ReferenceBatchSink<StructuredRecord, KEY_OUT, VAL_OUT> {
+  public static final String BUCKET_DES = "GCS Bucket used to store the data";
+  public static final String PROJECT_ID_DES = "Google Cloud Project ID with access to configured GCS buckets";
+  public static final String SERVICE_KEY_FILE_DES = "The Json_Key_File certificate file of the " +
+    "service account used for GCS access";
+  private static final String FILESYSTEM_PROPERTIES_DESCRIPTION = "A JSON string representing a map of properties " +
+    "needed for the distributed file system.";
+  private static final Gson GSON = new Gson();
+
+  private static final Type MAP_STRING_STRING_TYPE = new TypeToken<Map<String, String>>() { }.getType();
+
+  private final GCSSinkConfig config;
+
+  protected GCSBatchSink(GCSSinkConfig config) {
+    super(config);
+    this.config = config;
+    // update fileSystemProperties to include ProjectId and JsonKeyFile, so prepareRun can only set
+    // fileSystemProperties in configuration, and not deal with projectId and JsonKeyFile separately
+    // do not create file system properties if macros were provided unless in a test case
+    if (!this.config.containsMacro("fileSystemProperties") && !this.config.containsMacro("JsonKeyFile") &&
+      !this.config.containsMacro("ProjectId")) {
+      this.config.fileSystemProperties = updateFileSystemProperties(this.config.fileSystemProperties,
+                                                                    this.config.projectId, this.config.jsonKey);
+    }
+  }
+
+  @VisibleForTesting
+  GCSSinkConfig getConfig() {
+    return config;
+  }
+
+  @Override
+  public final void prepareRun(BatchSinkContext context) {
+    OutputFormatProvider outputFormatProvider = createOutputFormatProvider(context);
+    Map<String, String> outputConfig = new HashMap<>(outputFormatProvider.getOutputFormatConfiguration());
+    if (config.fileSystemProperties != null) {
```

Comment:
this is a bit odd. The subclass creates an OutputFormatProvider, then the parent takes the config map, adds some config settings, then creates another OutputFormatProvider.  Rather than this, you could pass the config map to the createOutputFormatProvider call, then use the OutputFormatProvider it returns directly.